### PR TITLE
Assume that `X = A.new` has type `A` in `# typed: true` files

### PR DESCRIPTION
**Pull request by jez**
### Motivation

"Why do I have to write an annotation for this" is a frequent question that new Sorbet users have. A lot of the time, the answer is for performance—Sorbet would rather not have to do a lot of expensive work to figure out what type to assume, and would rather the user just be explicit about what type the thing has.

However, in some cases this leads to a lot of redundancy.

In this change we take advantage of the fact that `A.new` **almost always** has type `A`, and use that to make an assumption. Sorbet will then check if that assumption is wrong and ask the user to correct it if necessary.

This puts a slightly larger burden on the user—if they are doing something where the result of calling `new` on a constant does not have type `A`, then they will get an error and be **forced** to add a type annotation (forced because I've chosen to make this behavior happen in `# typed: true` files too. This change would not make the user write any more type annotations than they otherwise would have had to had this been something that only fired in `# typed: strict` files). In any case, this error did not occur even once in Stripe's codebase so I think it should be quite rare.

### Test plan

See included automated tests.

### TODO

- [x] I have to figure out how to handle the "recv is a constant but not a class" error, because that currently leads to a terrible user experience.

### CODE REVIEW

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_MwRmZXzQtKy9SS
→ https://go/builds/bui_MwRmasWN5NbpV9

**Comment by jez** [comment]
> Pretty sure Stripe's codebase has instances of `A.new` overrides that don't actually return an instance of `A`. 😢 (I ran into these cases when optimizing `new` in the compiler.)

The idea is that if we limit this to `# typed: true` or even `# typed: strict` files, either (1) the typechecker tells us all the violations and we go fix them with a codemod before landing, or (2) we limit ourselves to `typed: strict` files, which must necessarily already have `T.let`, preventing the rewriter from firing.

**Comment by jez** [comment]
Update: it caused 35 failure when run at `# typed: true`. Going to try again at `typed: strict`.

**Comment by jez** [comment]
> Update: it caused 35 failure when run at `# typed: true`. Going to try again at `typed: strict`.

I misspoke, it actually only caused 21 errors at typed: true, and all of them were legitimate type errors.

I could go either way (do this at typed true or at typed strict). In any case, we can't land until we get a better error message that communicates what's going on.

File: infer/environment.cc
Comment (froydnj) [inline infer/environment.cc:1555]: I guess this offers a way forward to (maybe) easily dealing with #4837?  (Some of the issues from the last run in that PR -- which was admittedly a long time ago -- look not really autocorrectable, though.)

File: website/docs/error-reference.md
Comment (froydnj) [inline website/docs/error-reference.md:3919]: This paragraph strikes me as kind of weird given the preceding paragraphs explaining that Sorbet just tried to assume something and it turned out it was mistaken in that assumption.

e.g. what is "this way" in the first sentence referring to?  (It reads to me like it's referencing "...so that Sorbet doesn't have to assume anything"...but that referencing doesn't make sense in the context of this error message because clearly Sorbet is already making assumptions.)  Or what is "it doesn't have to go back and re-do a lot of work it has already done when it encounters a broken assumption" trying to get at?  I guess both of these are referencing the design principle that Sorbet does everything in a single pass, no backtracking or fixed points required, but I'm not sure how useful that is as user documentation.

Maybe this whole paragraph would just be better as a quick sentence pointing people at `why-type-annotations.md`?  But that also seems strange, since `why-type-annotations.md` is explaining, well, why type annotations are necessary and this error is coming from a place where type annotations weren't explicitly provided, but should be.

File: resolver/resolver.cc
Comment (froydnj) [inline resolver/resolver.cc:2818]: This is such a nice way of dealing with things.

File: rewriter/ConstantAssumeType.cc
Comment (froydnj) [inline rewriter/ConstantAssumeType.cc:32]: Unnecessary TODO now?

**Review by froydnj — APPROVED** [review]
I think this whole PR is a net win for users.  I'm not sure about the explanation in the docs.

File: infer/environment.cc
Comment (jez) [reply infer/environment.cc:1555]: That PR specifically is just asking "can we do the thing we already do with literals assigned to constants for instance variables as well" but yes we could then take it one step further and also do this transformation (i might take a look at this depending on how much time i have in the next few days)

File: website/docs/error-reference.md
Comment (jez) [reply website/docs/error-reference.md]: I took a stab at rewording this. I'm going to set the PR to auto-merge and we can rewrite this section of the docs in a future PR if you still think the follow up doesn't quite eliminate the weirdness you think it started out with.

**Pull request merged**
